Move left sidebar state to redux

noahtallen opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
Instead of prop drilling the sidebar state, this tries moving it to redux, and it also tries separating "navigation" from "inserter" state. cc @Copons ([I think you mentioned this here](https://github.com/WordPress/gutenberg/pull/25884#discussion_r502366419).)

## How has this been tested?
1. test that the changes in #25957 still work (i.e. the menu should go back to root when closing the panel)
2. You can test that this is working by clicking the "view in navigation" button which is displayed in the template dropdown:

![2020-10-13 17 55 05](https://user-images.githubusercontent.com/6265975/95931283-8b0ead00-0d7d-11eb-8f45-8f3cc8a54f2a.gif)


## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

noahtallen committed:
Move left sidebar state to redux

jeyip reviewed (commented):
This all makes sense to me. We could also use selectors to handle the "interdependent" behavior of the inserter and the navigation sidebar if we'd like. What do you think @Copons

david-szabo97 commented:
We already have a PR merged which saves the active menu into the store: https://github.com/WordPress/gutenberg/pull/25906 and we had a change by Jacopo here: https://github.com/WordPress/gutenberg/pull/25957

Changes look good and I see all the benefits of moving more state into the store. Could you do a rebase and make sure we only one have solution living the codebase? 😄

Copons commented on packages/edit-site/src/store/actions.js:
```diff
@@ -173,3 +173,33 @@ export function setNavigationPanelActiveMenu( menu ) {
+		type: 'SET_INSERTER_OPEN',
+		isOpen,
+	};
+}
```
Let's keep the namings consistent with the `edit-post` version:

https://github.com/WordPress/gutenberg/blob/0ee98c8f4ca2ab50f8d17e448b906021c787ac47/packages/edit-post/src/store/actions.js#L288-L299

https://github.com/WordPress/gutenberg/blob/0ee98c8f4ca2ab50f8d17e448b906021c787ac47/packages/edit-post/src/store/reducer.js#L241-L253

https://github.com/WordPress/gutenberg/blob/0ee98c8f4ca2ab50f8d17e448b906021c787ac47/packages/edit-post/src/store/selectors.js#L317-L326

noahtallen replied:
👍 done

Copons reviewed (changes requested):
The requested changes should be fairly straightforward, but overall this works well and I'm totally into this solution!
I've even proposed something like this in a comment in the other PR: https://github.com/WordPress/gutenberg/pull/25906#issuecomment-705035385 🙂 

As @david-szabo97 said: please rebase and replace the old state introduced in #25906 with this.
Also, please keep the logic introduced in #25957: when closing the navigation, we should revert the active menu to `root` (see [reasons](https://github.com/WordPress/gutenberg/pull/25957#issuecomment-706112103)).

> We could also use selectors to handle the "interdependent" behavior of the inserter and the navigation sidebar if we'd like.

@jeyip Could you please clarify this?
I think I'm misunderstanding what you're saying, but the reducer already makes inserter and navigation interdependent, by automatically closing one when we open the other. 🤔

Copons commented on packages/edit-site/src/components/template-details/index.js:
```diff
@@ -2,24 +2,35 @@
 			</p>
-
+			<Button onClick={ showTemplateInSidebar } isTertiary>
+				{ __( 'View in navigation.' ) }
```
I think this could use a translation context, or maybe even a reword (e.g. "View in navigation _sidebar_", which doesn't sound as good, but it's more clear). 🤔

noahtallen replied:
good idea. for what it's worth, I didn't spend any time on this button other than wiring it up, so it currently looks very out of place. mostly just for testing the redux changes

noahtallen commented:
> As @david-szabo97 said: please rebase and replace the old state introduced in #25906 with this.

I believe I had already incorporated `SET_NAVIGATION_PANEL_ACTIVE_MENU` into this PR, so I want to make sure I'm not missing something important or misunderstanding! I basically kept the existing action and selector, but added `action.menu` to the new `navigationPanel` store, wired in that action type, and also change the selector to point to the new state slice. So I think that aspect of things should be working.

> Also, please keep the logic introduced in #25957: when closing the navigation, we should revert the active menu to root

Additionally, I just moved the logic which sets the menu back to "root" when closing it from the navigation toggle component into the reducer. It's nice how much the reducer simplifies things!

noahtallen committed:
Add unit tests

jeyip commented:
> We could also use selectors to handle the "interdependent" behavior of the inserter and the navigation sidebar if we'd like.

> @jeyip Could you please clarify this?
> I think I'm misunderstanding what you're saying, but the reducer already makes inserter and navigation interdependent, by 
> automatically closing one when we open the other. 🤔

My mistake! You understood me perfectly @Copons.

1. The reducers _do_ already handle automatically closing one when we open the other.
2. I was talking about handling the business logic in selectors, which, in retrospect, is totally unnecessary (and somewhat confusing).

Copons reviewed (approved):
I've pushed a minor update that replaces hardcoded menu names with their corresponding constants.
This works well for me so I'm approving the PR, but please double check if I've accidentally caused regressions. 🙂

noahtallen commented:
@Copons thanks for that! Love the change. I'll do another round of testing and then merge.
